pkg/errors/user_errors.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errors
+
+import (
+	"errors"
+)
+
+type UserMessenger interface {
+	error
+	UserMessage() string
+}
+
+// UserError is an error type that be checked to be present in the
+// error chain. And if present, can be shown to the end users
+type UserError struct {
+	error
```

Comment:
Will there be cases where `UserError` needs to be a standalone error, without a parent `error`? Why not let `UserError` implements `errors` directly and the caller decide when/how they want to wrap `UserError` in a chain of errors?

Reply:
Imagine, you got an error:

```go
err := errors.New("somewhat")
```

And then you need an instance of `UserError` to present it to the user in a user-friendly manner. At the same time you want to preserve the behavior of `errors.Is`/`errors.As` for the error wrapped with `UserError`. How would you do that?

Reply:
I think updating `UserError` to implement `errors` is more idiomatic. There are probably other ways to go about this, but here is an example, where:

i. `UserError` can be treated like an independent `error`
i. The constructor uses `errors.As` to find and append existing `UserError`s to the new `UserError`, so that there is always one `UserError` to worry about during unwrapping

```diff
diff --git a/pkg/errors/user_errors.go b/pkg/errors/user_errors.go
index aa964207..5363b8fa 100644
--- a/pkg/errors/user_errors.go
+++ b/pkg/errors/user_errors.go
@@ -16,48 +16,51 @@ package errors
 
 import (
 	"errors"
+	"fmt"
+	"strings"
 )
 
-type UserMessenger interface {
-	error
-	UserMessage() string
-}
-
 // UserError is an error type that be checked to be present in the
 // error chain. And if present, can be shown to the end users
 type UserError struct {
-	error
 	msg string
 }
 
-func (ue *UserError) UserMessage() string {
+func (ue *UserError) Error() string {
 	return ue.msg
 }
 
-func (ue *UserError) Unwrap() error { return ue.error }
-
 // UserErrorWithMessage initializes the UserError with given error and a message.
 // Needs to accept error if we want to initialize the UserError in between the
 // error chain
-func UserErrorWithMessage(err error, msg string) UserMessenger {
+func WrapUserError(err error, msg string) error {
+	ue := &UserError{msg}
 	if err == nil {
-		err = errors.New(msg)
+		return ue
+	}
+
+	// append any existing UserErrors in err to msg, so there is always one
+	// msg to unwrap.
+	var u *UserError
+	if errors.As(err, &u) {
+		ue.msg += fmt.Sprintf("\n%s", u.Error())
 	}
-	return &UserError{err, msg}
+
+	return fmt.Errorf("%w %s", ue, err)
 }
 
 // UserMessagesInError gets us messages of all the `UserError`s that are wrapped
 // in the passed error's chain
-func UserMessagesInError(err error) []string {
+func UnwrapUserError(err error) []string {
 	userMessages := []string{}
 	if err == nil {
 		return nil
 	}
-	for err != nil {
-		if e, ok := err.(*UserError); ok {
-			userMessages = append(userMessages, e.UserMessage())
-		}
-		err = errors.Unwrap(err)
+
+	var ue *UserError
+	if errors.As(err, &ue) {
+		userMessages = strings.Split(ue.Error(), "\n")
 	}
+
 	return userMessages
 }
diff --git a/pkg/errors/user_errors_test.go b/pkg/errors/user_errors_test.go
index 5ab64162..91ce297e 100644
--- a/pkg/errors/user_errors_test.go
+++ b/pkg/errors/user_errors_test.go
@@ -32,13 +32,13 @@ func (u *UserErrorSuite) TestUserMessagesInError(c *C) {
 	err1 := errors.New("error 1")
 	err2 := errors.Wrap(err1, "error2")
 
-	ue1 := UserErrorWithMessage(err2, expectedUserErrors[1])
+	ue1 := WrapUserError(err2, expectedUserErrors[1])
 	err3 := errors.Wrap(ue1, "errro 3")
 
-	ue2 := UserErrorWithMessage(err3, expectedUserErrors[0])
+	ue2 := WrapUserError(err3, expectedUserErrors[0])
 	err4 := errors.Wrap(ue2, "error 4")
 
-	c.Assert(UserMessagesInError(err4), DeepEquals, expectedUserErrors)
+	c.Assert(UnwrapUserError(err4), DeepEquals, expectedUserErrors)
 }
 
 // The case where there are no `UserError`s in error chain
@@ -48,9 +48,9 @@ func (u *UserErrorSuite) TestUserMessagesInErrorWithoutUserError(c *C) {
 	err3 := errors.Wrap(err2, "error 3")
 	err4 := errors.Wrap(err3, "error 4")
 
-	c.Assert(len(UserMessagesInError(err4)), Equals, 0)
+	c.Assert(len(UnwrapUserError(err4)), Equals, 0)
 }
 
 func (u *UserErrorSuite) TestUserMessagesInErrorNilError(c *C) {
-	c.Assert(UserMessagesInError(nil), IsNil)
+	c.Assert(UnwrapUserError(nil), IsNil)
 }
```

Reply:
This suggestion doesn't change how the caller will call the exported functions, per the unit tests.

Reply:
Consider your proposal:

```go
var u *UserError
if errors.As(err, &u) {
    ue.msg += fmt.Sprintf("\n%s", u.Error())
}
```

As there is no loop anymore, what will happen to the potential further underlying user messages?

Also, I'm not sure that gluing/splitting the errors by "\n" is a good choice.